packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -386,6 +317,106 @@ library LibProving {
         });
     }
 
+    // TODO(brecht): can we write a circuit for this function?
+    function _proveAnchorForValidBlock(
+        TaikoData.Config memory config,
+        AddressResolver resolver,
+        IProofVerifier proofVerifier,
+        Evidence memory evidence,
+        bytes calldata anchorTx,
+        bytes calldata anchorReceipt
+    ) private view {
+        // Check anchor tx is valid
+        LibTxDecoder.Tx memory _tx = LibTxDecoder.decodeTx(
+            config.chainId,
+            anchorTx
+        );
+        if (_tx.txType != 0) revert L1_ANCHOR_TYPE();
+        if (_tx.destination != resolver.resolve(config.chainId, "taiko", false))
+            revert L1_ANCHOR_DEST();
+        if (_tx.gasLimit != config.anchorTxGasLimit)
+            revert L1_ANCHOR_GAS_LIMIT();
+
+        // Check anchor tx's signature is valid and deterministic
+        _validateAnchorTxSignature(config.chainId, _tx);
+
+        // Check anchor tx's calldata is valid
+        if (
+            !LibBytesUtils.equal(
+                _tx.data,
+                bytes.concat(
+                    ANCHOR_TX_SELECTOR,
+                    bytes32(evidence.meta.l1Height),
+                    evidence.meta.l1Hash
+                )
+            )
+        ) revert L1_ANCHOR_CALLDATA();
+
+        // Check anchor tx is the 1st tx in the block
+        uint zkProofsPerBlock = config.zkProofsPerBlock;
+        if (
+            !proofVerifier.verifyMKP({
+                key: LibRLPWriter.writeUint(0),
+                value: anchorTx,
+                proof: evidence.proofs[zkProofsPerBlock],
+                root: evidence.header.transactionsRoot
+            })
+        ) revert L1_ANCHOR_TX_PROOF();
+
+        // Check anchor tx does not throw
+        LibReceiptDecoder.Receipt memory receipt = LibReceiptDecoder
+            .decodeReceipt(anchorReceipt);
+        if (receipt.status != 1) revert L1_ANCHOR_RECEIPT_STATUS();
+
+        if (
+            !proofVerifier.verifyMKP({
+                key: LibRLPWriter.writeUint(0),
+                value: anchorReceipt,
+                proof: evidence.proofs[zkProofsPerBlock + 1],
+                root: evidence.header.receiptsRoot
+            })
+        ) revert L1_ANCHOR_RECEIPT_PROOF();
+    }
+
+    // TODO(brecht): can we write a circuit for this function?
```

Comment:
I don't think it will be worth it do this in the circuit. Hopefully we can do away with invalid blocks totally by having the proof on the proposer side.  But also because proposers will never propose invalid blocks anyway so it doesn't matter how expensive this is (as long as the gas cost is < Ethereum block gas limit).